_posts/2024-10-24-autorebalancing-on-scaling.md

```diff
@@ -0,0 +1,152 @@
+date: 2024-10-24
+author: paolo_patierno
+---
+
+Scaling an Apache Kafka cluster up or down can help meet various goals and deliver multiple benefits.
+Scaling up by adding new brokers enables the cluster to handle a higher data load across clients.
+Scaling down by removing brokers can reduce costs and energy consumption when demand is lower.
+But scaling the cluster makes more sense if it's done automatically, for example by leveraging the HPA (Horizontal Pod Autoscaler) within Kubernetes.
+After scaling up, the issue arises from the cluster being unbalanced, with new brokers left empty.
+This requires rebalancing topic partitions to equally spread the load across all brokers, old and new.
+On the other hand, scaling down may not be possible if the brokers to be removed host topic partitions, unless we are willing to accept offline replicas and sacrifice High Availability (HA).
+In both scenarios, rebalancing the cluster by using the Cruise Control integration within Strimzi is the solution.
+Doing it automatically makes the process even easier and this is what we are going to talk about in this blog post.
+
+<!--more-->
+
+### How scaling and rebalancing is done today
+
+When scaling up a Kafka cluster, newly added brokers don't get any partitions for the topics hosted on the other brokers.
+The new brokers remain empty and will receive partitions for newly created topics based on Apache Kafka's distribution mechanism.
+Quite often, this is not what you want because you are scaling up in order to spread the load across more brokers and getting better performance.
+Currently, the only way to achieve this with Strimzi is by manually rebalancing the cluster using the `KafkaRebalance` custom resource in `add-brokers` mode, where you list the added brokers and specify the goals to be met by Cruise Control.
+This way, some of the already existing topic partitions are moved from the old brokers to the new ones, making the cluster more balanced.
+
+When scaling down, the operation to remove brokers could be blocked by the Cluster Operator because those brokers are hosting topic partitions and the immediate shutdown will cause offline replicas and loss of HA.
+In order to move forward, before scaling down, it is possible to run a manual rebalancing by using a `KafkaRebalance` custom resource with the `remove-brokers` mode (listing the brokers to remove) and the goals for that.
+This way, Cruise Control moves partitions off the brokers to be removed so that they become empty and can be scaled down.
+
+For more details about how to do this, please refer to the official [documention](https://strimzi.io/docs/operators/latest/deploying#cruise-control-concepts-str) related to the Cruise Control integration within Strimzi.
+
+While both the above approaches work, the process is still actually manual and done in two separate steps:
+
+* scale the cluster up, then run a rebalancing.
+* run a rebalancing, then scale the cluster down.
```

Comment:
```suggestion
* scale the cluster up, then run an `add-brokers` rebalancing.
* run a `remove-brokers` rebalancing, then scale the cluster down.
```

Reply:
I made the change in a slightly different way to mention "mode".